lalrpop/src/lr1/interpreter.rs

```diff
@@ -0,0 +1,273 @@
+        let error_type = self.types.error_type();
+        let parse_error_type = self.parse_error_type();
+
+        let (type_parameters, parameters);
+
+        if self.grammar.intern_token.is_some() {
+            // if we are generating the tokenizer, we just need the
+            // input, and that has already been added as one of the
+            // user parameters
+            type_parameters = vec![];
+            parameters = vec![];
+        } else {
+            // otherwise, we need an iterator of type `TOKENS`
+            let mut user_type_parameters = String::new();
+            for type_parameter in &self.grammar.type_parameters {
+                user_type_parameters.push_str(&format!("{}, ", type_parameter));
+            }
+            type_parameters = vec![
+                format!("{}TOKEN: {}ToTriple<{}Error={}>",
+                        self.prefix, self.prefix, user_type_parameters, error_type),
+                format!("{}TOKENS: IntoIterator<Item={}TOKEN>", self.prefix, self.prefix)];
+            parameters = vec![format!("{}tokens: {}TOKENS", self.prefix, self.prefix)];
+        }
+
+        try!(self.out.write_pub_fn_header(
+            self.grammar,
+            format!("parse_{}", self.user_start_symbol),
+            type_parameters,
+            parameters,
+            format!("Result<{}, {}>",
+                    self.types.nonterminal_type(self.start_symbol),
+                    parse_error_type),
+            vec![]));
+        rust!(self.out, "{{");
+
+        if self.grammar.intern_token.is_some() {
+            // if we are generating the tokenizer, create a matcher as our input iterator
+            rust!(self.out, "let mut {}tokens = super::{}intern_token::{}Matcher::new(input);",
+                  self.prefix, self.prefix, self.prefix);
+        } else {
+            // otherwise, convert one from the `IntoIterator`
+            // supplied, using the `ToTriple` trait which inserts
+            // errors/locations etc if none are given
+            rust!(self.out, "let {}tokens = {}tokens.into_iter();", self.prefix, self.prefix);
+            rust!(self.out, "let mut {}tokens = {}tokens.map(|t| {}ToTriple::to_triple(t));",
+                  self.prefix, self.prefix, self.prefix);
+        }
+
+
+        //rust!(self.out, "    let machine = Machine::new(&actions, &gotos, &productions);");
+        rust!(self.out, "}}");
+        Ok(())
+    }
+
+    fn write_terminal_to_index_fn(&mut self) -> io::Result<()> {
+        try!(self.out.write_pub_fn_header(
+            self.grammar,
+            "terminal_to_index".to_owned(),
+            vec![],
+            vec![format!("token: {}", self.grammar.types.terminal_token_type())],
+            "usize".to_owned(),
+            vec![]));
+        rust!(self.out, "{{");
+        rust!(self.out, "match token {{");
+
+        for (&k, v) in &self.grammar.terminal_bits {
+            // TODO same as in ascent.rs
```

Comment:
the ascent.rs code has the notion of `match_terminal_pattern`, which is a pattern that just decides what token you are looking at, and `consume_terminal`, which also extracts out the data. The latter includes bindings. It seems like you are emitting a "consume" pattern here but not using the `pattern_names` anywhere. Nonetheless, I suspect that consuming is right: you probably want to match the lookahead and produce `(index, value)` where `value` is some variant of the "data stack" enum corresponding to this kind of terminal (and hence containing the data for the terminal).